bin/appup_test.sh

```diff
@@ -0,0 +1,180 @@
+#!/usr/bin/env bash
+set -euo pipefail
+BASEDIR=$(dirname $(realpath "$0"))
+
+## @DOC
+## The script is used to test the appup.src file with following steps,
+## 1) It builds all release tar files of *current* vsn and old vsns that they are defined in ${supported_rel_vsns}
+## 2) It untar all the tar files to one tmp dir then build the *relup* file from there. see make_relup()
+## 3) It then append the relup file to the tar file of current release, see make_relup()
+## 4) It then test the old vsn upgrade and downgrade, see test_relup()
+
+# GLOBAL
+app=emqtt
+supported_rel_vsns="1.4.6"
+
+die()
+{
+    echo "$1"
+    exit 1;
+}
+
+build_and_save_tar() {
+    dest_dir="$1"
+
+    if rebar3 plugins list |grep relup_helper; then
+        rebar3 as emqtt_relup_test do relup_helper gen_appups,tar;
+    else
+        rebar3 as emqtt_relup_test do tar
+    fi
+
+    mv _build/emqtt_relup_test/rel/emqtt/emqtt-*.tar.gz "${dest_dir}"
+}
+
+build_legacy() {
+    dest_dir="$1"
+    for vsn in ${supported_rel_vsns};
+    do
+        echo "building rel tar for $vsn"
+        vsn_dir="$dest_dir/$app-$vsn"
+        # FIXME, this is temp repo for test
+        git clone https://github.com/emqx/emqtt.git -b "$vsn" --recursive --depth 1 "$vsn_dir"
+        pushd ./
+        cd ${vsn_dir}
+        build_and_save_tar "$dest_dir";
+        popd
+    done
+}
+
+untar_all_pkgs() {
+    local dir="$1/"
+    local appdir="$dir/$app"
+    mkdir -p "$appdir"
+    for f in ${dir}/*.tar.gz;
+    do
+        tar zxf "$f" -C "$appdir";
+    done
+}
+
+prepare_releases() {
+    local dest_dir="$1"
+    build_and_save_tar "$dest_dir"
+    build_legacy "$dest_dir"
+    untar_all_pkgs "$dest_dir"
+}
+
+erl_eval() {
+    local node_cmd="$1"
+    local cmd="$2"
+    local expected_res="$3"
+    [ ! -f $node_cmd ] && die "$node_cmd not found"
+    res=$($node_cmd eval "$2")
+    if [[ $expected_res != $res  ]]; then
+       die "Failed: eval: $cmd\n but returns $res"
+    fi
+}
+
+test_relup() {
+    local tar_dir="$1"
+    local target_vsn="$2"
+
+    for vsn in ${supported_rel_vsns};
+    do
+        echo "unpack"
+        appdir="${tar_dir}/${vsn}/"
+        rm --preserve-root -rf "${appdir}/"
+        mkdir -p ${appdir}
+        tar zxf "$tar_dir/$app-$vsn.tar.gz" -C "$appdir"
+
+        ##
+        ## Start Old Version of EMQTT
+        ##
+        echo "starting $vsn"
+        appscript="${appdir}/bin/emqtt"
+        trap "$appscript stop" EXIT
+        $appscript daemon -- -mode interactive
+        $appscript ping
+        $appscript versions
+
+        ##
+        ## Deploy NEW Target Version
+        ##
+        echo "deploy $target_vsn"
+        cp "$tar_dir/$app-$target_vsn.tar.gz" "$appdir/releases/"
+        $appscript versions
+
+        $appscript eval 'spawn_link(fun() -> process_flag(trap_exit, false), {ok, Pid}=emqtt:start_link(), ets:insert(ac_tab,{{application_master, emqtt}, Pid}), true=register(test_client1, Pid), receive stop -> ok end end).'
```

Comment:
For my own information: if the process is spawned here with `spawn_link`, to what is it linked to?  I expected that the process would immediately die after the spawn, since the eval has finished.  And indeed I couldn't make it live past the eval in emqx.  :thinking: 

Maybe this should be just `spawn`?

Reply:
Good point, my original vsn was trap exit true, but I change it to false to verify if bash could return non-zero value if commands are wrong. Apparently, I forget to change it back.